### PR TITLE
Add implicit field read steps when promoted fields are accessed

### PR DESCRIPTION
**Pull request by smowton**
For previous discussion see https://github.com/github/codeql-go/pull/492

Performance results for this are now worst-case +50s / +7% (Kubernetes).

Having invested a lot of effort trying to optimise this, and given it is a correctness fix, I'm inclined to merge this now and pursue performance fixups down the line.

### CODE REVIEW

**Review by sauyon — APPROVED** [review]
LGTM. It's a bit unfortunate that we have to be so verbose about a lot of this stuff to satisfy the compiler.

**Comment by owen-mc** [comment]
I'm part-way through a review - should be done soon

File: ql/src/semmle/go/controlflow/IR.qll [hunk 1]
```diff
@@ -253,21 +257,76 @@ module IR {
     )
   }
 
+  /**
+   * Gets the effective base of a selector, index or slice expression, taking implicit dereferences
+   * and implicit field reads into account.
+   *
+   * For a selector expression `b.f`, this could be the implicit dereference `*b`, or the implicit
+   * field access `b.Embedded.f` if the field `f` is promoted from an embedded type `Embedded`, or
```
Comment (owen-mc) [inline ql/src/semmle/go/controlflow/IR.qll:265]:
```suggestion
   * field access `b.Embedded` if the field `f` is promoted from an embedded type `Embedded`, or
```

File: ql/src/semmle/go/controlflow/IR.qll [hunk 2]
```diff
@@ -253,21 +257,76 @@ module IR {
...
+   * a combination of both, or simply `b` if neither case applies.
```
Comment (owen-mc) [inline ql/src/semmle/go/controlflow/IR.qll:266]: In some sense it could also be `*(b.Embedded)`, although that and `b.Embedded` are represented by the same `Instruction`.

File: ql/src/semmle/go/controlflow/IR.qll [hunk 3]
```diff
@@ -253,21 +257,76 @@ module IR {
...
+   */
+  private Instruction selectorBase(Expr e) {
```
Comment (owen-mc) [inline ql/src/semmle/go/controlflow/IR.qll:268]: I find the ordering of the methods a bit weird, that this one is between the two methods it could call. Maybe move it below `promotedFieldSelectorBase`?

File: ql/src/semmle/go/controlflow/IR.qll [hunk 4]
```diff
@@ -253,21 +257,76 @@ module IR {
...
+    result = promotedFieldSelectorBase(e, _)
```
Comment (owen-mc) [inline ql/src/semmle/go/controlflow/IR.qll:269]: ❗ I'm really surprised you can get away with putting `_` here. I'd have expected you to have to get the `Field` that `e.getSelector()` refers to. What's to stop it from using a different field from the struct type of the base? `promotedFieldSelectorBase` doesn't constrain `field` any more than that. (Note that our tests don't have any structs with more than one non-promoted field.)

File: ql/src/semmle/go/controlflow/IR.qll [hunk 5]
```diff
@@ -277,12 +336,67 @@ module IR {
    * misclassified as field reads.
    */
   class FieldReadInstruction extends ComponentReadInstruction {
-    override SelectorExpr e;
+    SelectorExpr e;
+    Field field;
+
+    FieldReadInstruction() {
+      e = this.(EvalInstruction).getExpr() and
+      field.getAReference() = this.(EvalInstruction).getExpr().(SelectorExpr).getSelector()
```
Comment (owen-mc) [inline ql/src/semmle/go/controlflow/IR.qll:344]: We could shorten this to
```suggestion
      field.getAReference() = e.getSelector()
```

File: ql/src/semmle/go/controlflow/IR.qll [hunk 6]
```diff
@@ -277,12 +336,67 @@ module IR {
...
+    override predicate readsField(Instruction base, Field f) {
+      base = selectorBase(e, f) and f = field
+    }
+  }
+
+  /**
+   * An IR instruction for an implicit field read as part of reading a
+   * promoted field.
+   *
+   * If that field has a pointer type then this instruction also represents an
```
Comment (owen-mc) [inline ql/src/semmle/go/controlflow/IR.qll:362]: I just realised that at first glance, "that field" appears to be referring to the promoted field at the end of the previous sentence.
```suggestion
   * If the field that is being implicitly read has a pointer type then this instruction also represents an
```

**Review by owen-mc — COMMENTED** [review]
Most of my comments are very minor. The non-minor ones I've marked with ❗ so they stand out.

File: ql/test/library-tests/semmle/go/dataflow/PromotedFields/main.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func sink(s string) {}
+
+type Inner struct {
+	field string
+}
+
+type Middle struct {
+	Inner
```
Comment (owen-mc) [inline ql/test/library-tests/semmle/go/dataflow/PromotedFields/main.go:14]: We should make one of the embedded fields have a pointer type, to check that we're dealing with that correctly.
```suggestion
	*Inner
```

Comment (owen-mc) [reply ql/test/library-tests/semmle/go/dataflow/PromotedFields/main.go:14]: Currently we do not support this: https://github.com/github/codeql-go-team/issues/270

**Comment by owen-mc** [comment]
Superseded by https://github.com/github/codeql-go/pull/492
